Schedule featured scenes

Carifio24 opened the pull request:
This PR adds an initial implementation of allowing scenes to be scheduled to be featured at a certain future time, as well a general feature queue to be used on days when there are no features scheduled. Here, at least for now, "featured" means that it will float to the top of the home timeline.

This functionality requires two additional collections in the database:
* `features` contains the set of features. Each document contains a scene ID and a datetime - see `MongoSceneFeature`
* `featureQueue` should contain a single document with the structure:
```
{
  queue: true,
  scene_ids: [<ObjectIDs>]  // This list contains the IDs of the queue of scenes
}
```

The `queue` field doesn't really do anything, but the various methods here do query for it. The idea is that if another document accidentally gets put into this collection, we at least have a decent chance of being able to find the queue successfully. Obviously this is not an ideal setup as it doesn't really leverage the power of having a database - it's just a list (though to be fair, Mongo does provide some operations for manipulating arrays). I've seen implementations of linked lists in Mongo that might be worth looking into, but for now our list shouldn't be long enough that this will cause problems.

Additionally, this contains some routes for manipulating features and the queue, and modifies the timeline setup to account for these additions.

pkgw commented on features.ts:
```diff
@@ -0,0 +1,390 @@
+  });
+}
+
+export async function tryPopFromFeatureQueue(state: State): Promise<ModifyResult<MongoSceneFeatureQueue>> {
```
Is this function going to do what we want? As I understand it the `$pop` operator removes an item from the affected list, and that item just ... disappears. I believe the return value of the `findOneAndUpdate` is the affected document, after the pop has been applied. But for our "pop"-type operations, we care about what that item was. Am I missing something here?

Carifio24 replied:
We discussed this the other day, but just to record for posterity: the [`findOneAndUpdate`](https://mongodb.github.io/node-mongodb-native/6.3/classes/Collection.html#findOneAndUpdate) method actually returns the original, pre-update document, unless you tell it otherwise. (In a Node context, that's done via setting the `returnDocument` field in the query [options](https://mongodb.github.io/node-mongodb-native/6.3/interfaces/FindOneAndUpdateOptions.html#returnDocument))

pkgw commented on features.ts:
```diff
@@ -0,0 +1,390 @@
+  state.app.post(
+    "/features/queue/pop",
+    async (req: JwtRequest, res: Response) => {
+      const result = await tryPopFromFeatureQueue(state);
```
As per my comment on the `tryPopFromFeatureQueue` function, I am not sure if this endpoint will function as intended. It seems like we need to fetch the queue doc, save the 0'th ID, and then update the doc to remove that ID from the list.

(This would have busted semantics if multiple people are trying to update the doc at once. I'm not worried about that corner case right now, but if Mongo/Cosmos offers a true "pop" that can remove an item from the list *and return it to us all in one atomic operation*, that would be better.)

Carifio24 replied:
It's worth noting that the operation actually is atomic (per the docs [here](https://mongodb.github.io/node-mongodb-native/6.3/classes/Collection.html#findOneAndUpdate)). Though we are missing any kind of write-lock handling, so there is still some future work to be done.

pkgw commented on features.ts:
```diff
@@ -0,0 +1,390 @@
+      newQueue.push(objectId);
+      const updateResult = await state.featureQueue.updateOne(
+        { queue: true },
+        { scene_ids: newQueue }
```
Is there an operator that can delete an item from a list? That would potentially have more robust semantics in the off chance that someone is trying to update the queue from elsewhere simultaneously. But, again, we really don't expect that kind of contention for this API, and it's low-stakes if anything goes wrong, so it's not a big deal if this is the best available way to implement the deletion.

Carifio24 replied:
Yep, we can use `$pull` - I used `$push` for adding an item, but completely forgot about `$pull` for some reason.

pkgw commented on superuser.ts:
```diff
@@ -194,8 +195,30 @@ export function initializeSuperuserEndpoints(state: State) {
+          const result = await tryPopFromFeatureQueue(state);
+          const queueDoc = result.value;
+          initialSceneID = queueDoc?.scene_ids[0] ?? null;
+        }
```
I feel like this logic would be better located in a little function in `features.ts`, so that this code can just write something like

```
initialSceneID = await getCurrentFeaturedScene(state);
```

or something like that. The other file seems like a better place to put that kind of logic, which depends on relatively detailed properties of the feature system.

pkgw replied:
Reupping this comment.

pkgw reviewed (commented):
Looking good! I do have a number of comments, below.

pkgw commented:
Marking as a draft since there are security issues to be resolved before merging.

Carifio24 commented on features.ts:
```diff
@@ -0,0 +1,390 @@
+export async function getFeaturesForDate(state: State, date: Date): Promise<FindCursor<WithId<MongoSceneFeature>>> {
+  const day = timeStrippedDate(date);
+  const nextDay = new Date(day);
+  nextDay.setDate(nextDay.getDate() + 1);
```
This actually does work for the last day of the month - the JavaScript `Date` object is [smart enough](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/setDate#Description) to move into the next month if the number of days exceeds the month's count.

I've removed the `async` label.

Carifio24 commented:
@pkgw I think all of the issues have been addressed now. Most of the changes were pretty straightforward, but a couple that I feel like are at least worth mentioning:

* In `superuser.ts`, I refactored the superuser verification functionality out of the endpoint-creation method so that we could reuse it for the features.
* I've added a logging middleware for the features endpoints using `@azure/logger` that displays some basic information about what the incoming request looks like.
* I've also added some functionality that allows passing in a "magic key" in a request header. This is only used for the features endpoints, where either the magic key or being a superuser is enough for a request to be handled.

pkgw reviewed (commented):
OK, I think we're almost there, but when it comes to security-related stuff I want to be as thorough as possible.

The issue that I see with the current code is that it's being more open about permissions than it needs to be. There is a very specific list of API calls that will be made autonomously by the frontend server (I think literally just one?), and *only* those calls are the ones where the frontend magic key should count as sufficient authorization.

Likewise if there are API calls that only the frontend will make, but a superuser won't realistically make directly, the superuser authorization shouldn't be accepted. We want to minimize the open surface area wherever we can. We can always loosen targeted APIs if/when we realize that they need to be opened up a bit.

Also, the term "Constellations key" is very vague. I think it would be better to be more specific there, especially since (as per the principles above) best practice would be to minimize the privileges associated with any such key — we're not going to want a single magic key that's accepted for a dozen different functionalities. So, maybe something like "frontend autonomous key" or something, since in my mind that's what's distinctive about this one: we need some way for the frontend to issue its own privileged requests, as opposed to doing things on behalf of a logged-in user.

Also, the environment variable should be documented in the README.

pkgw commented:
I made a few final changes, some of which are important:

- `/features/queue` didn't have any auth middleware — in principle this information could be public, but it I made it superuser-only to start
- The check for a frontend key was asking for an HTTP header named `CX_FRONTEND_AUTONOMOUS_KEY`, which would get normalized but is not a valid HTTP header name. I changed the check to look for a header of the form `Authorization: CXInternal {key}`.
- In the key-or-superuser auth middleware, if auth is denied, you don't want to leak any information about the check. (Since we're open-source it's essentially moot, but it's all about consistently doing the right thing here.)
